### ISSUE TITLE
RISCV Isel: Cannot select: 0x55a041ad6990: nxv2i32 = RISCVISD::VZEXT_VL 0x55a04197a350, 0x55a041ad6840, Register:i64 $x0 

### ISSUE THREAD

**Issue by hiraditya** [llvm/llvm-project#83654]
Error
```
1.      <eof> parser at end of file   
2.      Code generation
3.      Running pass 'Function Pass Manager' on module '/usr/local/foobar/out/llvm-project/libcxx/src/locale.cpp'.
4.      Running pass 'RISC-V DAG->DAG Pattern Instruction Selection' on function '@_ZNSt3__114__scan_keywordB8ne180000INS_19istreambuf_iteratorIcNS_11char_t
raitsIcEEEEPKNS_12basic_stringIcS3_NS_9allocatorIcEEEENS_5ctypeIcEEEET0_RT_SE_SD_SD_RKT1_Rjb'
 #0 0x000055a03d6787f8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/local/foobar/out/stage2-install/bin/clang++
+0x31c47f8)
 #1 0x000055a03d67660e llvm::sys::RunSignalHandlers() (/usr/local/foobar/out/stage2-install/bin/clang+++0x31c260e)
 #2 0x000055a03d677cde llvm::sys::CleanupOnSignal(unsigned long) (/usr/local/foobar/out/stage2-install/bin/clang+++0x31c3cde
)
 #3 0x000055a03d602f3e (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) CrashRecoveryContext.cpp:0:0
 #4 0x000055a03d602efb (/usr/local/foobar/out/stage2-install/bin/clang+++0x314eefb)
 #5 0x000055a03d673787 llvm::sys::Process::Exit(int, bool) (/usr/local/foobar/out/stage2-install/bin/clang+++0x31bf787)
 #6 0x000055a03c3a58a3 (/usr/local/foobar/out/stage2-install/bin/clang+++0x1ef18a3)
 #7 0x000055a03d605c92 llvm::report_fatal_error(llvm::Twine const&, bool) (/usr/local/foobar/out/stage2-install/bin/clang+++
0x3151c92)
 #8 0x000055a03e42fa2c (/usr/local/foobar/out/stage2-install/bin/clang+++0x3f7ba2c)
 #9 0x000055a03e42efe5 (/usr/local/foobar/out/stage2-install/bin/clang+++0x3f7afe5)
#10 0x000055a03c8b5dce llvm::RISCVDAGToDAGISel::Select(llvm::SDNode*) RISCVISelDAGToDAG.cpp:0:0
#11 0x000055a03e42670f llvm::SelectionDAGISel::DoInstructionSelection() (/usr/local/foobar/out/stage2-install/bin/clang+++0x
3f7270f)
#12 0x000055a03e425e79 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/usr/local/foobar/out/stage2-install/bin/clang+++0x3f71e
79)
#13 0x000055a03e425011 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/usr/local/foobar/out/stage2-ins
tall/bin/clang+++0x3f71011)
#14 0x000055a03e4230f4 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/usr/local/foobar/out/stage2-in
stall/bin/clang+++0x3f6f0f4)
#15 0x000055a03cf21bf7 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/local/foobar/out/stage2-install/bin/c
lang+++0x2a6dbf7)
...

clang++: warning: argument unused during compilation: '-stdlib=libc++' [-Wunused-command-line-argument]
fatal error: error in backend: Cannot select: 0x55a041ad6990: nxv2i32 = RISCVISD::VZEXT_VL 0x55a04197a350, 0x55a041ad6840, Register:i64 $x0
  0x55a04197a350: nxv2i1 = setcc 0x55a04197a4a0, 0x55a0433e8d10, seteq:ch
    0x55a04197a4a0: nxv2i64 = vselect 0x55a041979fd0, 0x55a0433e8fb0, 0x55a04197a6d0
      0x55a041979fd0: nxv2i1 = setcc 0x55a0433e9170, 0x55a041979da0, seteq:ch
        0x55a0433e9170: nxv2i8 = and 0x55a041979e80, 0x55a04197a820
          0x55a041979e80: nxv2i8,ch = llvm.riscv.vlse<(load unknown-size from %ir.68, align 1)> 0x55a044f22ee0, TargetConstant:i64<10129>, undef:nxv2i8, 0x5
5a04197a890, Constant:i64<24>, Register:i64 $x0
            0x55a04197a0b0: i64 = TargetConstant<10129>
            0x55a04197a580: nxv2i8 = undef
            0x55a04197a890: i64,ch = CopyFromReg 0x55a044f22ee0, Register:i64 %14
              0x55a04197a900: i64 = Register %14
            0x55a04197a040: i64 = Constant<24>
            0x55a041ad67d0: i64 = Register $x0
          0x55a04197a820: nxv2i8 = RISCVISD::VMV_V_X_VL undef:nxv2i8, Constant:i64<1>, Register:i64 $x0
            0x55a04197a580: nxv2i8 = undef
            0x55a0433e8ed0: i64 = Constant<1>
            0x55a041ad67d0: i64 = Register $x0
        0x55a041979da0: nxv2i8 = RISCVISD::VMV_V_X_VL undef:nxv2i8, Constant:i64<0>, Register:i64 $x0
          0x55a04197a580: nxv2i8 = undef
          0x55a0433e8760: i64 = Constant<0>
          0x55a041ad67d0: i64 = Register $x0
      0x55a0433e8fb0: nxv2i64 = zero_extend nneg 0x55a0433e88b0
        0x55a0433e88b0: nxv2i8 = srl 0x55a041979e80, 0x55a04197a820
          0x55a041979e80: nxv2i8,ch = llvm.riscv.vlse<(load unknown-size from %ir.68, align 1)> 0x55a044f22ee0, TargetConstant:i64<10129>, undef:nxv2i8, 0x55a04197a890, Constant:i64<24>, Register:i64 $x0
            0x55a04197a0b0: i64 = TargetConstant<10129>
            0x55a04197a580: nxv2i8 = undef
            0x55a04197a890: i64,ch = CopyFromReg 0x55a044f22ee0, Register:i64 %14
              0x55a04197a900: i64 = Register %14
            0x55a04197a040: i64 = Constant<24>
            0x55a041ad67d0: i64 = Register $x0
          0x55a04197a820: nxv2i8 = RISCVISD::VMV_V_X_VL undef:nxv2i8, Constant:i64<1>, Register:i64 $x0
            0x55a04197a580: nxv2i8 = undef
            0x55a0433e8ed0: i64 = Constant<1>
            0x55a041ad67d0: i64 = Register $x0
      0x55a04197a6d0: nxv2i64,ch = llvm.riscv.vluxei<(load unknown-size, align 8, !alias.scope !49)> 0x55a044f22ee0, TargetConstant:i64<10173>, undef:nxv2i64, 0x55a0433e9250, 0x55a04197a2e0, Register:i64 $x0
        0x55a0433e8df0: i64 = TargetConstant<10173>
        0x55a0433e8840: nxv2i64 = undef
        0x55a0433e9250: i64 = add 0x55a04197a890, Constant:i64<8>
          0x55a04197a890: i64,ch = CopyFromReg 0x55a044f22ee0, Register:i64 %14
            0x55a04197a900: i64 = Register %14
          0x55a0433e85a0: i64 = Constant<8>
        0x55a04197a2e0: nxv2i64,ch = CopyFromReg 0x55a044f22ee0, Register:nxv2i64 %11
          0x55a04197a7b0: nxv2i64 = Register %11
        0x55a041ad67d0: i64 = Register $x0
    0x55a0433e8d10: nxv2i64 = RISCVISD::VMV_V_X_VL undef:nxv2i64, Constant:i64<0>, Register:i64 $x0
      0x55a0433e8840: nxv2i64 = undef
      0x55a0433e8760: i64 = Constant<0>
      0x55a041ad67d0: i64 = Register $x0
  0x55a041ad6840: nxv2i1 = RISCVISD::VMSET_VL Register:i64 $x0
    0x55a041ad67d0: i64 = Register $x0
  0x55a041ad67d0: i64 = Register $x0
In function: _ZNSt3__114__scan_keywordB8ne180000INS_19istreambuf_iteratorIcNS_11char_traitsIcEEEEPKNS_12basic_stringIcS3_NS_9allocatorIcEEEENS_5ctypeIcEEEET0_RT_SE_SD_SD_RKT1_Rjb
```

**Comment by hiraditya** [comment]
Repro:
preprocessed file and the script.

[locale-680c83.sh.txt](https://github.com/llvm/llvm-project/files/14467959/locale-680c83.sh.txt)
[locale-680c83.cpp.txt](https://github.com/llvm/llvm-project/files/14467960/locale-680c83.cpp.txt)

**Comment by topperc** [comment]
@sun-jacobi ca you take a look. Might be related to ba81477e9cdb5384a10c700ee247562f0f6c03e6

**Comment by sun-jacobi** [comment]
The operands for `RISCVISD::VZEXT_VL` should not be `nxv2i1`.

I will try to fix it soon.

**Comment by lukel97** [comment]
I can't reproduce the crash on main @ 3f7aa042b657671319f994ad3fb7c3eb79a6fe00. I don't think it's related to https://github.com/llvm/llvm-project/commit/ba81477e9cdb5384a10c700ee247562f0f6c03e6 or combineBinOp_VLToVWBinOp_VL, otherwise we would see a VWADD_VL node or similar in the graph?

**Comment by lukel97** [comment]
The revision of LLVM used in the attached test script seems to be 3c92011b60 based off of https://android.googlesource.com/toolchain/llvm-project/+/d8003a456d14a3deb8054cdaa529ffbf02d9b262 (I'm not familiar with the android toolchain, I could be reading this wrong).

@hiraditya are you able to reproduce this with a more recent revision?

**Comment by topperc** [comment]
There was this revert on Jan 2 for a bad VZEXT_VL

```
commit 4e347b4e38b95bc455d0e620e11ac58fc0172a94
Author: Craig Topper <craig.topper@sifive.com>
Date:   Tue Jan 2 19:38:49 2024

    Revert "[RISCV][ISel] Combine scalable vector add/sub/mul with zero/sign extension (#72340)"
    
    This reverts most of commit 5b155aea0e529b7b5c807e189fef6ea5cd5faec9.
    I have left the new test file, but regenerated the checks.
    
    This causes failures in our downstream testing. The input types
    to the extends need to be checked so we don't create RISCVISD::VZEXT_VL
    with illegal or unsupported input type.
```

**Comment by lukel97** [comment]
I've bisected it back to 5b155aea0e52, so it does look like it's from combineBinOp_VLToVWBinOp_VL.

It was fixed in the revert 4e347b4e38b95bc455d0e620e11ac58fc0172a94 and remained fixed after the recommit ba81477e9cdb5384a10c700ee247562f0f6c03e6, so presumably it was the same issue that was mentioned in https://github.com/llvm/llvm-project/pull/72340#issuecomment-1874810998

I think this is fixed now.

**Comment by hiraditya** [comment]
Doesn't repro in latest compiler. Thanks for verifying.
